dali/kernels/imgproc/convolution/convolution_cpu.h

```diff
@@ -0,0 +1,362 @@
+        float acc = 0;
+        for (int k = 0; k < window_size; k++) {
+          int x = boundary::idx_reflect_101(x0 + k, axis_size);
+          acc += in_axis[x * channels + c] * window[k];
+        }
+        out_axis[xout * channels + c] = ConvertSat<Out>(acc * scale);
+      }
+    }
+  }
+}
+
+template <int axis, bool has_channels, int max_lanes, typename Out, typename In, typename W,
+          int ndim>
+void ConvolveInplaceAxisLoop(Out* out, const In* in, const W* window,
+                             const TensorShape<ndim>& shape, const TensorShape<ndim>& strides,
+                             int diameter, int64_t offset, In* input_window_buffer, W scale,
+                             int num_lanes) {
+  auto axis_stride = strides[axis];
+  auto axis_size = shape[axis];
+  int radius = (diameter - 1) / 2;
+  // offset <- start of current axis
+  auto* out_ptr = out + offset;
+  auto* in_ptr = in + offset;
+
+  CyclicWindowWrapper<In, max_lanes> input_window(input_window_buffer, diameter, num_lanes);
+  int in_idx = -radius, out_idx = 0;
+  for (in_idx = -radius; in_idx < 0; in_idx++) {
+    load_pixel_with_border(input_window, in_ptr, in_idx, axis_stride, axis_size);
+  }
+  // we load the window without the last element
+  for (; in_idx < radius && in_idx < axis_size; in_idx++) {
+    load_pixel_no_border(input_window, in_ptr, in_idx, axis_stride);
+  }
+  // if we already went out of the window, fill the rest
+  for (; in_idx < radius; in_idx++) {
+    load_pixel_with_border(input_window, in_ptr, in_idx, axis_stride, axis_size);
+  }
+  // if we're still in the window, use version without border, till the in_idx goes out
+  for (; out_idx < axis_size && in_idx < axis_size; out_idx++, in_idx++) {
+    // we load last element of the input window corresponding to the out_idx
+    load_pixel_no_border(input_window, in_ptr, in_idx, axis_stride);
+    input_window.CalculateDot(out_ptr + out_idx * axis_stride, window, scale);
+    // remove one element, to make space for next out_idx and in_idx
+    input_window.PopElement();
+  }
+  // finish the rest of output
+  for (; out_idx < axis_size; out_idx++, in_idx++) {
+    // To process in-place, we need to pick the vales back from the CyclicBuffer,
+    // as it may happen that we already stored the element.u
+    reload_pixel_with_border(input_window, in_ptr, in_idx, out_idx, axis_stride, axis_size, radius);
+    input_window.CalculateDot(out_ptr + out_idx * axis_stride, window, scale);
+    input_window.PopElement();
+  }
+}
+
+template <int axis, bool has_channels, int max_lanes, typename Out, typename In, typename W,
+          int ndim>
+void ConvolveInplaceOuterLoop(Out* out, const In* in, const W* window,
+                              const TensorShape<ndim>& shape, const TensorShape<ndim>& strides,
+                              int diameter, In* input_window_buffer, W scale = 1) {
+  int64_t outer_elements = volume(&shape[0], &shape[axis]);
```

Comment:
can you add a static assert to check that axis is within the valid range? Unless you are sure it is because it's checked before